kvserver: filter rangefeed events based on OmitInRangefeeds flag

miraradeva opened the pull request:
OmitInRangefeeds is a transaction attibute that, when set to true, indicates that rangefeeds should ignore all writes of the transaction. It is set on all transactions for sessions that have the session variable `disable_changefeed_replication`. The flag is also populated on the MVCCValueHeader of any transactional write whose trasaction has it set to true. This change enforces the actual filtering on the rangefeed server side for both current rangefeed events and the rangefeed catchup iterator.

If a user sets the `disable_changefeed_replication` session variable and also runs statements that write to system tables (e.g. set a cluster setting), the rangefeed events corresponding to those writes will be dropped because the use transaction is used for the writes; this is undesirable because many internal system components depend on these events (e.g. the cluster settings rangefeed cache). To prevent this, we also add a flag on the rangefeed consumer side to explicitly opt into filtering. All changefeeds will set this flag to true to utilize the `disable_changefeed_replication` feature, and all internal rangefeeds will not set the flag to ensure they receive all rangefeed events.

Part of: #114313

Release note: None

miraradeva committed:
kvserver: filter rangefeed events based on OmitInRangefeeds flag

OmitInRangefeeds is a transaction attibute that, when set to true,
indicates that rangefeeds should ignore all writes of the transaction.
It is set on all transactions for sessions that have the session
variable `disable_changefeed_replication`. The flag is also populated
on the MVCCValueHeader of any transactional write whose trasaction has
it set to true. This change enforces the actual filtering on the
rangefeed server side for both current rangefeed events and the
rangefeed catchup iterator.

If a user sets the `disable_changefeed_replication` session variable and
also runs statements that write to system tables (e.g. set a cluster
setting), the rangefeed events corresponding to those writes will be
dropped because the use transaction is used for the writes; this is
undesirable because many internal system components depend on these
events (e.g. the cluster settings rangefeed cache). To prevent this, we
also add a flag on the rangefeed consumer side to explicitly opt into
filtering. All changefeeds will set this flag to true to utilize the
`disable_changefeed_replication` feature, and all internal rangefeeds
will not set the flag to ensure they receive all rangefeed events.

Part of: #114313

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/116065)
<!-- Reviewable:end -->

miraradeva commented:
TODO: the rangefeed consumer changes will come in a separate commit shortly.

miraradeva committed:
kvcoord,kvfeed: opt changefeeds into rangefeed filtering

The rangefeeds server exposes a flag (`WithFiltering`) that lets
rangefeed clients opt into rangefeed filtering. Currently, filtering
includes only the `OmitInRangefeeds` flag, which in turn is controlled
by the `disable_changefeed_replication` session variable. When
filtering is enabled, transactional writes with `OmitInRangefeeds` set
to true are not included in the rangefeed.

Changefeeds should always opt into filtering to leverage the
`disable_changefeed_replication` feature. All other rangefeed clients
should never opt into filtering. In particular, rangefeeds for system
tables should not use filtering to ensure they observe all system table
rangefeed events.

Part of: #114313

Release note: None

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-:-NlPaiBw1ff61SWds1KT:bnfp4nl)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jayshrivastava, @miraradeva, and @nvanbenschoten)

___
*[`pkg/kv/kvpb/api.proto` line 3120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-NlP_ZMg9hO6enm_Cd11:-NlP_ZMhEUqAg4pAwr8_:b-9pspzw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb81a016b36f37965aba8173f628ebc8963fc64/pkg/kv/kvpb/api.proto#L3120)):*
> ```proto
>   // OmitInRangefeeds = true, the write will not be emitted on the rangefeed.
>   // WithFiltering should NOT be set for system-table rangefeeds.
>   bool with_filtering = 7;
> ```

I like this choice of a safe default :thumbsup: 
___
*[`pkg/kv/kvserver/replica_rangefeed.go` line 761 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-NlP_owP4PTBfQ6Gxx87:-NlP_owP4PTBfQ6Gxx88:b-m1qlpn) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb81a016b36f37965aba8173f628ebc8963fc64/pkg/kv/kvserver/replica_rangefeed.go#L761)):*
> ```Go
> 		// event out of the rangefeed of any registrations that opted into filtering.
> 		if o, ok := op.GetValue().(*enginepb.MVCCCommitIntentOp); ok {
> 			if vh.OmitInRangefeeds {
> ```

[nit] We should check `vh.OmitInRangeFeeds` first
___
*[`pkg/kv/kvserver/rangefeed/processor.go` line 806 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-NlPa8Ki8JoPt1XJklGv:-NlPa8Kj7gUDL4SwE7Rw:b89epie) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb81a016b36f37965aba8173f628ebc8963fc64/pkg/kv/kvserver/rangefeed/processor.go#L806)):*
> ```Go
> 		case *enginepb.MVCCWriteValueOp:
> 			// Publish the new value directly.
> 			p.publishValue(ctx, t.Key, t.Timestamp, t.Value, t.PrevValue, false /* omitInRangefeeds */, alloc)
> ```

Why not `t.OmitInRangeFeeds` here? (maybe add a comment explaining)
___
*[`pkg/storage/enginepb/mvcc3.proto` line 302 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-NlPaXbE-ENP7Fflmn3d:-NlPaXbE-ENP7Fflmn3e:bkxd1aw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb81a016b36f37965aba8173f628ebc8963fc64/pkg/storage/enginepb/mvcc3.proto#L302)):*
> ```proto
>   // OmitInRangefeeds is populated from the MVCCValueHeader of the corresponding
>   // write. It is only relevant for transactional writes.
>   bool omit_in_rangefeeds = 6;
> ```

[nit] also explain what it does (maybe copy the description - or part of it - from MVCCValueHeader)


<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/kv/kvserver/rangefeed/scheduled_processor.go:
```diff
@@ -753,7 +755,7 @@ func (p *ScheduledProcessor) publishDeleteRange(
 		Timestamp: timestamp,
 	})
-	p.reg.PublishToOverlapping(ctx, span, &event, alloc)
+	p.reg.PublishToOverlapping(ctx, span, &event, false /* omitInRangefeeds */, alloc)
```
might be nicer to add an enum/constant to avoid passing these "false /* ... */" things....

nvb reviewed (commented):
> Reviewed 15 of 15 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116065)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jayshrivastava and @miraradeva)

___
*[`pkg/kv/kvserver/replica_rangefeed.go` line 760 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-NlRF6417LfZvsk8Hy6a:-NlRF6417LfZvsk8Hy6b:badyjw5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb81a016b36f37965aba8173f628ebc8963fc64/pkg/kv/kvserver/replica_rangefeed.go#L760)):*
> ```Go
> 		// later, in the registry's PublishToOverlapping, to potentially filter this
> 		// event out of the rangefeed of any registrations that opted into filtering.
> 		if o, ok := op.GetValue().(*enginepb.MVCCCommitIntentOp); ok {
> ```

Don't we also want to attach this flag on `MVCCWriteValueOp`s?
___
*[`pkg/kv/kvserver/replica_rangefeed.go` line 760 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-NlRH0wJ5nTpI3LJN1fP:-NlRH0wJ5nTpI3LJN1fQ:b-epp9h2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dd49d1f9b6109b9bb3ea276bbf78ae3eda70f89/pkg/kv/kvserver/replica_rangefeed.go#L760)):*
> ```Go
> 		// later, in the registry's PublishToOverlapping, to potentially filter this
> 		// event out of the rangefeed of any registrations that opted into filtering.
> 		if o, ok := op.GetValue().(*enginepb.MVCCCommitIntentOp); ok {
> ```

My comment about `MVCCWriteValueOp` raises the question: are we assigning this field in the MVCCValueHeader for 1-phase commit writes that evaluate under the `evaluate1PC` path?
___
*[`pkg/kv/kvserver/replica_rangefeed_test.go` line 408 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-NlRGYna1h2r0y8vge4U:-NlRGYna1h2r0y8vge4V:bnq56oa) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb81a016b36f37965aba8173f628ebc8963fc64/pkg/kv/kvserver/replica_rangefeed_test.go#L408)):*
> ```Go
> 	dArgs := delRangeArgs(roachpb.Key("c"), roachpb.Key("d"), true /* useRangeTombstone */)
> 	_, err = kv.SendWrappedWith(ctx, db, kvpb.Header{Timestamp: ts8}, dArgs)
> 	if err != nil {
> ```

nit throughout: these tests are all using the `require` library. We should keep using it here.


<!-- Sent from Reviewable.io -->

miraradeva reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116065)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jayshrivastava, @nvanbenschoten, and @RaduBerinde)

___
*[`pkg/kv/kvserver/replica_rangefeed.go` line 760 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-NlRF6417LfZvsk8Hy6a:-NlTqQvP9wrJPoUl1T-t:bbujkt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb81a016b36f37965aba8173f628ebc8963fc64/pkg/kv/kvserver/replica_rangefeed.go#L760)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Don't we also want to attach this flag on `MVCCWriteValueOp`s?
</blockquote></details>

Yes, looks like we do, for 1PC writes.
___
*[`pkg/kv/kvserver/replica_rangefeed.go` line 760 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-NlRH0wJ5nTpI3LJN1fP:-NlTqWYt8CpBIqE4r5jZ:b-a32he7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dd49d1f9b6109b9bb3ea276bbf78ae3eda70f89/pkg/kv/kvserver/replica_rangefeed.go#L760)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

My comment about `MVCCWriteValueOp` raises the question: are we assigning this field in the MVCCValueHeader for 1-phase commit writes that evaluate under the `evaluate1PC` path?
</blockquote></details>

No, we're not, and we should be. Thanks for catching this. 

The only place I can think of where we can do this is in `evaluate1PC` where we still have the txn. We can populate `OmitInRangefeeds` directly in the response's `LogicalOps`. I don't see a way to do it in `mvccPutInternal` to where it will show up in the `MVCCValueHeader`; we don't know anything about 1PC there. 
___
*[`pkg/kv/kvserver/rangefeed/processor.go` line 806 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-NlPa8Ki8JoPt1XJklGv:-NlTrJqqCIV2QOhuygUU:b-n317kn) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fb81a016b36f37965aba8173f628ebc8963fc64/pkg/kv/kvserver/rangefeed/processor.go#L806)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Why not `t.OmitInRangeFeeds` here? (maybe add a comment explaining)
</blockquote></details>

Yeah, we need to do it here too.


<!-- Sent from Reviewable.io -->

miraradeva reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116065)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jayshrivastava, @miretskiy, @nvanbenschoten, and @RaduBerinde)

___
*[`pkg/kv/kvserver/replica_rangefeed.go` line 760 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-NlRH0wJ5nTpI3LJN1fP:-NlUJl-x7zT3gO9NqU8p:b9di638) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dd49d1f9b6109b9bb3ea276bbf78ae3eda70f89/pkg/kv/kvserver/replica_rangefeed.go#L760)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

No, we're not, and we should be. Thanks for catching this. 

The only place I can think of where we can do this is in `evaluate1PC` where we still have the txn. We can populate `OmitInRangefeeds` directly in the response's `LogicalOps`. I don't see a way to do it in `mvccPutInternal` to where it will show up in the `MVCCValueHeader`; we don't know anything about 1PC there.
</blockquote></details>

I did this and it works for 1PC ongoing writes but not for the catchup iterator. There we really need the `OmitInRangefeeds` flag to be in the `MVCCValueHeader` and I don't know how to get it in there.
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 758 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116065#gh-1423197861:-NlUK651B6HomvdHBpcW:b5we1hf) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dd49d1f9b6109b9bb3ea276bbf78ae3eda70f89/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L758)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

might be nicer to add an enum/constant to avoid passing these "false /* ... */" things....
</blockquote></details>

In most places I tried to keep it `withDiff` and `withFiltering` consistent (and they account for most of these arguments). I agree they can both be refactored but maybe not in this PR.


<!-- Sent from Reviewable.io -->

miraradeva reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116065)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jayshrivastava, @miretskiy, @nvanbenschoten, and @RaduBerinde)

___
*[`pkg/kv/kvserver/replica_rangefeed.go` line 760 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-NlRH0wJ5nTpI3LJN1fP:-NlUtttA0t8qQbjFeFH2:baf7r47) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dd49d1f9b6109b9bb3ea276bbf78ae3eda70f89/pkg/kv/kvserver/replica_rangefeed.go#L760)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

I did this and it works for 1PC ongoing writes but not for the catchup iterator. There we really need the `OmitInRangefeeds` flag to be in the `MVCCValueHeader` and I don't know how to get it in there.
</blockquote></details>

I guess the way to pass `OmitInRangefeeds` to the `MVCCValueHeader` for 1PC is to include it as a new flag in the `BatchRequest` header, and then pass it to `mvccPutInternal` via `MVCCWriteOptions`. I'm happy to try it out if we think it's not too risky modifying the `BatchRequest` proto.


<!-- Sent from Reviewable.io -->

miraradeva reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116065)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jayshrivastava, @miretskiy, @nvanbenschoten, and @RaduBerinde)

___
*[`pkg/kv/kvserver/replica_rangefeed.go` line 760 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-NlRH0wJ5nTpI3LJN1fP:-Nld0rORDxgEDLHqO1ij:b6qrxtf) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dd49d1f9b6109b9bb3ea276bbf78ae3eda70f89/pkg/kv/kvserver/replica_rangefeed.go#L760)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

I guess the way to pass `OmitInRangefeeds` to the `MVCCValueHeader` for 1PC is to include it as a new flag in the `BatchRequest` header, and then pass it to `mvccPutInternal` via `MVCCWriteOptions`. I'm happy to try it out if we think it's not too risky modifying the `BatchRequest` proto.
</blockquote></details>

After discussing with Nathan, decided not to modify the request header, and instead pass the flag through a few evaluation functions and down to the MVCC level as `CommandArgs`. It simplified the handling of the two sources of the `OmitInRangefeeds` flag (one from the transaction proto and one from 1PC), and doesn't involve a proto change.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
> Reviewed 5 of 13 files at r3, 21 of 21 files at r5, 5 of 5 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116065)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jayshrivastava, @miraradeva, @miretskiy, and @RaduBerinde)

___
*[`pkg/kv/kvserver/replica_rangefeed.go` line 761 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-NldoCLvCWDJ3oGPuDfP:-NldoCLvCWDJ3oGPuDfQ:b-y0hide) ([raw file](https://github.com/cockroachdb/cockroach/blob/436e5e606b4967e9499273b79804872eab4df20e/pkg/kv/kvserver/replica_rangefeed.go#L761)):*
> ```Go
> 		// PublishToOverlapping, to potentially filter this event out of the
> 		// rangefeed of any registrations that opted into filtering.
> 		if vh.OmitInRangefeeds {
> ```

We should only be calling `MVCCGetForKnownTimestampWithNoIntent` if `op` is one of the two types anyway. So maybe we can restructure this to avoid duplication. We could pull out a `var omitInRangefeedPtr *bool` up above, like we have for  `valPtr`. Then just `*omitInRangefeedPtr = vh.OmitInRangefeeds`.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-:-NldpTXh-ea3Ob5UqAUf:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @jayshrivastava, @miraradeva, @miretskiy, and @RaduBerinde)



<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 5 of 10 files at r7, 5 of 5 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116065#-:-NleM4R11d3tcGPlcc8Q:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @jayshrivastava, @miretskiy, and @RaduBerinde)



<!-- Sent from Reviewable.io -->

miraradeva commented:
bors r=nvanbenschoten
